File: core/kora-app-annotation-processor/src/test/java/io/koraframework/kora/app/annotation/processor/ConflictResolutionTest.java
package io.koraframework.kora.app.annotation.processor;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class ConflictResolutionTest extends AbstractKoraAppTest {
    @Test
    public void testMultipleComponentCandidates() {
        var result = compile(List.of(new KoraAppProcessor()), """
            public interface TestInterface {}
            """, """
            public class TestImpl1 implements TestInterface {}
            """, """
            public class TestImpl2 implements TestInterface {}
            """, """
            @KoraApp
            public interface ExampleApplication {
                @Root
                default String root(TestInterface t) {return "";}
                default TestImpl1 testImpl1() { return new TestImpl1(); }
                default TestImpl2 testImpl2() { return new TestImpl2(); }
            }
            """);
        assertThat(result.isFailed()).isTrue();
    }

    @Test
    public void testDefaultComponentOverride() throws ClassNotFoundException {
        var draw = compile("""
            public interface TestInterface {}
            """, """
            public class TestImpl1 implements TestInterface {}
            """, """
            public class TestImpl2 implements TestInterface {}
            """, """
            @KoraApp
            public interface ExampleApplication {
                @Root
                default String root(TestInterface t) {return "";}
                @DefaultComponent
                default TestImpl1 testImpl1() { return new TestImpl1(); }
                default TestImpl2 testImpl2() { return new TestImpl2(); }
            }
            """);
        assertThat(draw.getNodes()).hasSize(2);
        var g = draw.init();
        assertThat(g.get(draw.getNodes().get(0))).isInstanceOf(this.compileResult.loadClass("TestImpl2"));
    }

    @Test
    public void testDefaultComponentTemplateOverride() throws ClassNotFoundException {
        var draw = compile("""
            public interface TestInterface <T> {}
            """, """
            public class TestImpl1 <T> implements TestInterface<T> {}
            """, """
            public class TestImpl2 <T> implements TestInterface<T> {}
            """, """
            @KoraApp
            public interface ExampleApplication {
                @Root
                default String root(TestInterface<String> t) {return "";}

                @DefaultComponent
                default <T> TestImpl1<T> testImpl1() { return new TestImpl1<>(); }
                default <T> TestImpl2<T> testImpl2() { return new TestImpl2<>(); }
            }
            """);
        assertThat(draw.getNodes()).hasSize(2);
        var g = draw.init();
        assertThat(g.get(draw.getNodes().get(0))).isInstanceOf(this.compileResult.loadClass("TestImpl2"));
    }

    @Test
    public void testDefaultComponentAnnotatedComponentUsedWhenNoOverride() throws ClassNotFoundException {
        var draw = compile("""
            public interface TestInterface {}
            """, """
            @Component
            @DefaultComponent
            public class DefaultTestImpl implements TestInterface {}
            """, """
            @KoraApp
            public interface ExampleApplication {
                @Root
                default Object root(TestInterface t) {return t;}
            }
            """);
        assertThat(draw.getNodes()).hasSize(2);

        var defaultImpl = this.compileResult.loadClass("DefaultTestImpl");
        var g = draw.init();
        assertThat(draw.getNodes())
            .extracting(g::get)
            .anySatisfy(value -> assertThat(value).isInstanceOf(defaultImpl));
    }

    @Test
    public void testDefaultComponentAnnotatedComponentOverrideByComponentClass() throws ClassNotFoundException {
        var draw = compile("""
            public interface TestInterface {}
            """, """
            @Component
            @DefaultComponent
            public class DefaultTestImpl implements TestInterface {}
            """, """
            @Component
            public class OverrideTestImpl implements TestInterface {}
            """, """
            @KoraApp
            public interface ExampleApplication {
                @Root
                default Object root(TestInterface t) {return t;}
            }
            """);
        assertThat(draw.getNodes()).hasSize(2);

        var defaultImpl = this.compileResult.loadClass("DefaultTestImpl");
        var overrideImpl = this.compileResult.loadClass("OverrideTestImpl");
        var g = draw.init();
        assertThat(draw.getNodes())
            .extracting(g::get)
            .anySatisfy(value -> assertThat(value).isInstanceOf(overrideImpl))
            .noneSatisfy(value -> assertThat(value).isInstanceOf(defaultImpl));
    }

    @Test
    public void testDefaultComponentAnnotatedComponentOverrideByFactoryMethod() throws ClassNotFoundException {
        var draw = compile("""
            public interface TestInterface {}
            """, """
            @Component
            @DefaultComponent
            public class DefaultTestImpl implements TestInterface {}
            """, """
            public class OverrideTestImpl implements TestInterface {}
            """, """
            @KoraApp
            public interface ExampleApplication {
                @Root
                default Object root(TestInterface t) {return t;}

                default OverrideTestImpl testImpl() { return new OverrideTestImpl(); }
            }
            """);
        assertThat(draw.getNodes()).hasSize(2);

        var defaultImpl = this.compileResult.loadClass("DefaultTestImpl");
        var overrideImpl = this.compileResult.loadClass("OverrideTestImpl");
        var g = draw.init();
        assertThat(draw.getNodes())
            .extracting(g::get)
            .anySatisfy(value -> assertThat(value).isInstanceOf(overrideImpl))
            .noneSatisfy(value -> assertThat(value).isInstanceOf(defaultImpl));
    }

    @Test
    public void testDefaultComponentAnnotatedComponentOverrideBySameTagOnly() throws ClassNotFoundException {
        var draw = compile("""
            public interface TestInterface {}
            """, """
            public final class DefaultTag {}
            """, """
            public final class OverrideTag {}
            """, """
            @Component
            @DefaultComponent
            @Tag(DefaultTag.class)
            public class DefaultTestImpl implements TestInterface {}
            """, """
            @Component
            @Tag(OverrideTag.class)
            public class OverrideTestImpl implements TestInterface {}
            """, """
            @KoraApp
            public interface ExampleApplication {
                @Root
                default Object root(@Tag(DefaultTag.class) TestInterface t) {return t;}
            }
            """);
        assertThat(draw.getNodes()).hasSize(2);

        var defaultImpl = this.compileResult.loadClass("DefaultTestImpl");
        var overrideImpl = this.compileResult.loadClass("OverrideTestImpl");
        var g = draw.init();
        assertThat(draw.getNodes())
            .extracting(g::get)
            .anySatisfy(value -> assertThat(value).isInstanceOf(defaultImpl))
            .noneSatisfy(value -> assertThat(value).isInstanceOf(overrideImpl));
    }

    @Test
    public void testDefaultComponentAnnotatedComponentOverrideByComponentClassWithSameTag() throws ClassNotFoundException {
        var draw = compile("""
            public interface TestInterface {}
            """, """
            public final class TestTag {}
            """, """
            @Component
            @DefaultComponent
            @Tag(TestTag.class)
            public class DefaultTestImpl implements TestInterface {}
            """, """
            @Component
            @Tag(TestTag.class)
            public class OverrideTestImpl implements TestInterface {}
            """, """
            @KoraApp
            public interface ExampleApplication {
                @Root
                default Object root(@Tag(TestTag.class) TestInterface t) {return t;}
            }
            """);
        assertThat(draw.getNodes()).hasSize(2);

        var defaultImpl = this.compileResult.loadClass("DefaultTestImpl");
        var overrideImpl = this.compileResult.loadClass("OverrideTestImpl");
        var g = draw.init();
        assertThat(draw.getNodes())
            .extracting(g::get)
            .anySatisfy(value -> assertThat(value).isInstanceOf(overrideImpl))
            .noneSatisfy(value -> assertThat(value).isInstanceOf(defaultImpl));
    }

    @Test
    public void testDefaultComponentAnnotatedComponentOverrideByConditionalComponent() throws ClassNotFoundException {
        var draw = compile("""
            public interface TestInterface {}
            """, """
            @Component
            @DefaultComponent
            public class DefaultTestImpl implements TestInterface {}
            """, """
            @Component
            @Conditional(tag = io.koraframework.kora.app.annotation.processor.ConditionalComponentTest.MatchesCondition.class)
            public class OverrideTestImpl implements TestInterface {}
            """, """
            @KoraApp
            public interface ExampleApplication {
                @Root
                default Object root(TestInterface t) {return t;}

                @Tag(io.koraframework.kora.app.annotation.processor.ConditionalComponentTest.MatchesCondition.class)
                default GraphCondition matches() { return new io.koraframework.kora.app.annotation.processor.ConditionalComponentTest.MatchesCondition(); }
            }
            """);
        assertThat(draw.getNodes()).hasSize(3);

        var defaultImpl = this.compileResult.loadClass("DefaultTestImpl");
        var overrideImpl = this.compileResult.loadClass("OverrideTestImpl");
        var g = draw.init();
        assertThat(draw.getNodes())
            .extracting(g::get)
            .anySatisfy(value -> assertThat(value).isInstanceOf(overrideImpl))
            .noneSatisfy(value -> assertThat(value).isInstanceOf(defaultImpl));
    }

}

File: core/kora-app-symbol-processor/src/test/kotlin/io/koraframework/kora/app/ksp/ConflictResolutionTest.kt
package io.koraframework.kora.app.ksp

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ConflictResolutionTest : AbstractKoraAppProcessorTest() {
    @Test
    fun testMultipleComponentSameType() {
        compile0(listOf(KoraAppProcessorProvider()),
            """
            interface TestInterface            
            """.trimIndent(),
            """
                        class TestImpl1 : TestInterface {}
                        """.trimIndent(), """
                                    class TestImpl2 : TestInterface {}
                                    """.trimIndent(), """
                                                @KoraApp
                                                interface ExampleApplication {
                                                    @Root
                                                    fun root(t: TestInterface) = ""
                                                    fun testImpl1() = TestImpl1()
                                                    fun testImpl2() = TestImpl2()
                                                }
                                                
                                                """.trimIndent()
        )

        compileResult.assertFailure()
    }

    @Test
    fun testDefaultComponentOverride() {
        compile0(listOf(KoraAppProcessorProvider()),
            """
            interface TestInterface            
            """.trimIndent(),
            """
                        class TestImpl1 : TestInterface {}
                        """.trimIndent(), """
                                    class TestImpl2 : TestInterface {}
                                    """.trimIndent(), """
                                                @KoraApp
                                                interface ExampleApplication {
                                                    @Root
                                                    fun root(t: TestInterface) = ""
                                                
                                                    fun testImpl1() = TestImpl1()
                                    
                                                    @DefaultComponent
                                                    fun testImpl2() = TestImpl2()
                                                }
                                                """.trimIndent()
        )

        compileResult.assertSuccess()
    }

    @Test
    fun testDefaultComponentTemplateOverride() {
        compile0(listOf(KoraAppProcessorProvider()),
            """
            interface TestInterface <T>
            """.trimIndent(),
            """
                        class TestImpl1 <T> : TestInterface <T> {}
                        """.trimIndent(), """
                                    class TestImpl2 <T> : TestInterface <T> {}
                                    """.trimIndent(), """
                                                @KoraApp
                                                interface ExampleApplication {
                                                    @Root
                                                    fun root(t: TestInterface<String>) = ""
                                                
                                                    fun <T> testImpl1() = TestImpl1<T>()
                                    
                                                    @DefaultComponent
                                                    fun <T> testImpl2() = TestImpl2<T>()
                                                }
                                                """.trimIndent()
        )

        compileResult.assertSuccess()
    }

    @Test
    fun testDefaultComponentAnnotatedComponentUsedWhenNoOverride() {
        val draw = compile(
            """
            interface TestInterface
            """,
            """
            @Component
            @DefaultComponent
            class DefaultTestImpl : TestInterface
            """,
            """
            @KoraApp
            interface ExampleApplication {
                @Root
                fun root(t: TestInterface): Any = t
            }
            """
        )
        assertThat(draw.nodes).hasSize(2)

        val defaultImpl = loadClass("DefaultTestImpl")
        val graph = draw.init()
        val values = draw.nodes.map { graph.get(it) }
        assertThat(values).anyMatch { defaultImpl.isInstance(it) }
    }

    @Test
    fun testDefaultComponentAnnotatedComponentOverrideByComponentClass() {
        val draw = compile(
            """
            interface TestInterface
            """,
            """
            @Component
            @DefaultComponent
            class DefaultTestImpl : TestInterface
            """,
            """
            @Component
            class OverrideTestImpl : TestInterface
            """,
            """
            @KoraApp
            interface ExampleApplication {
                @Root
                fun root(t: TestInterface): Any = t
            }
            """
        )
        assertThat(draw.nodes).hasSize(2)

        val defaultImpl = loadClass("DefaultTestImpl")
        val overrideImpl = loadClass("OverrideTestImpl")
        val graph = draw.init()
        val values = draw.nodes.map { graph.get(it) }
        assertThat(values).anyMatch { overrideImpl.isInstance(it) }
        assertThat(values).noneMatch { defaultImpl.isInstance(it) }
    }

    @Test
    fun testDefaultComponentAnnotatedComponentOverrideByFactoryMethod() {
        val draw = compile(
            """
            interface TestInterface
            """,
            """
            @Component
            @DefaultComponent
            class DefaultTestImpl : TestInterface
            """,
            """
            class OverrideTestImpl : TestInterface
            """,
            """
            @KoraApp
            interface ExampleApplication {
                @Root
                fun root(t: TestInterface): Any = t

                fun testImpl() = OverrideTestImpl()
            }
            """
        )
        assertThat(draw.nodes).hasSize(2)

        val defaultImpl = loadClass("DefaultTestImpl")
        val overrideImpl = loadClass("OverrideTestImpl")
        val graph = draw.init()
        val values = draw.nodes.map { graph.get(it) }
        assertThat(values).anyMatch { overrideImpl.isInstance(it) }
        assertThat(values).noneMatch { defaultImpl.isInstance(it) }
    }

    @Test
    fun testDefaultComponentAnnotatedComponentOverrideBySameTagOnly() {
        val draw = compile(
            """
            interface TestInterface
            """,
            """
            class DefaultTag
            """,
            """
            class OverrideTag
            """,
            """
            @Component
            @DefaultComponent
            @Tag(DefaultTag::class)
            class DefaultTestImpl : TestInterface
            """,
            """
            @Component
            @Tag(OverrideTag::class)
            class OverrideTestImpl : TestInterface
            """,
            """
            @KoraApp
            interface ExampleApplication {
                @Root
                fun root(@Tag(DefaultTag::class) t: TestInterface): Any = t
            }
            """
        )
        assertThat(draw.nodes).hasSize(2)

        val defaultImpl = loadClass("DefaultTestImpl")
        val overrideImpl = loadClass("OverrideTestImpl")
        val graph = draw.init()
        val values = draw.nodes.map { graph.get(it) }
        assertThat(values).anyMatch { defaultImpl.isInstance(it) }
        assertThat(values).noneMatch { overrideImpl.isInstance(it) }
    }

    @Test
    fun testDefaultComponentAnnotatedComponentOverrideByComponentClassWithSameTag() {
        val draw = compile(
            """
            interface TestInterface
            """,
            """
            class TestTag
            """,
            """
            @Component
            @DefaultComponent
            @Tag(TestTag::class)
            class DefaultTestImpl : TestInterface
            """,
            """
            @Component
            @Tag(TestTag::class)
            class OverrideTestImpl : TestInterface
            """,
            """
            @KoraApp
            interface ExampleApplication {
                @Root
                fun root(@Tag(TestTag::class) t: TestInterface): Any = t
            }
            """
        )
        assertThat(draw.nodes).hasSize(2)

        val defaultImpl = loadClass("DefaultTestImpl")
        val overrideImpl = loadClass("OverrideTestImpl")
        val graph = draw.init()
        val values = draw.nodes.map { graph.get(it) }
        assertThat(values).anyMatch { overrideImpl.isInstance(it) }
        assertThat(values).noneMatch { defaultImpl.isInstance(it) }
    }

    @Test
    fun testDefaultComponentAnnotatedComponentOverrideByConditionalComponent() {
        val draw = compile(
            """
            interface TestInterface
            """,
            """
            @Component
            @DefaultComponent
            class DefaultTestImpl : TestInterface
            """,
            """
            @Component
            @Conditional(tag = io.koraframework.kora.app.ksp.ConditionalComponentTest.MatchesCondition::class)
            class OverrideTestImpl : TestInterface
            """,
            """
            @KoraApp
            interface ExampleApplication {
                @Root
                fun root(t: TestInterface): Any = t

                @Tag(io.koraframework.kora.app.ksp.ConditionalComponentTest.MatchesCondition::class)
                fun matches(): GraphCondition = io.koraframework.kora.app.ksp.ConditionalComponentTest.MatchesCondition()
            }
            """
        )
        assertThat(draw.nodes).hasSize(3)

        val defaultImpl = loadClass("DefaultTestImpl")
        val overrideImpl = loadClass("OverrideTestImpl")
        val graph = draw.init()
        val values = draw.nodes.map { graph.get(it) }
        assertThat(values).anyMatch { overrideImpl.isInstance(it) }
        assertThat(values).noneMatch { defaultImpl.isInstance(it) }
    }
}
